repose-aggregator/components/filters/openstack-identity-v3/src/main/scala/org/openrepose/filters/openstackidentityv3/OpenStackIdentityV3Handler.scala

```diff
@@ -59,43 +57,43 @@ class OpenStackIdentityV3Handler(identityConfig: OpenstackIdentityV3Config, iden
       region = Option(serviceEndpoint.getRegion))
   }
 
-  override def handleRequest(request: HttpServletRequest, response: ReadableHttpServletResponse): FilterDirector = {
-    val filterDirector: FilterDirector = new FilterDirectorImpl()
+  def handleRequest(request: HttpServletRequestWrapper, response: HttpServletResponse): FilterAction = {
+    var filterAction = FilterAction.PASS
 
     def delegateOrElse(responseCode: Int, message: String)(f: => Unit) {
       delegatingWithQuality match {
         case Some(quality) =>
           buildDelegationHeaders(responseCode, "openstack-identity-v3", message, quality) foreach { case (key, values) =>
-            filterDirector.requestHeaderManager.appendHeader(key, values: _*)
-            filterDirector.setFilterAction(FilterAction.PROCESS_RESPONSE)
-            filterDirector.setResponseStatusCode(200) // Note: The response status code must be set to a non-500 so that the request will be routed appropriately.
+            values foreach { value =>
+              request.addHeader(key, value)
+            }
+            filterAction = FilterAction.PROCESS_RESPONSE
+            response.setStatus(HttpServletResponse.SC_OK) // Note: The response status code must be set to a non-500 so that the request will be routed appropriately.
           }
         case None =>
           f
       }
     }
 
-    def authServiceOverLimit(e: IdentityServiceOverLimitException, filterDirector: FilterDirector): Unit = {
-      filterDirector.setResponseStatusCode(HttpServletResponse.SC_SERVICE_UNAVAILABLE)
+    def authServiceOverLimit(e: IdentityServiceOverLimitException): Unit = {
+      response.setStatus(HttpServletResponse.SC_SERVICE_UNAVAILABLE)
       var retry = e.getRetryAfter
       if (retry == null) {
         val retryCalendar = new GregorianCalendar
         retryCalendar.add(Calendar.SECOND, 5)
         retry = new HttpDate(retryCalendar.getTime).toRFC1123
       }
-      filterDirector.responseHeaderManager.appendHeader(HttpHeaders.RETRY_AFTER, retry)
+      response.addHeader(HttpHeaders.RETRY_AFTER, retry)
     }
 
     // Check if the request URI is whitelisted and pass it along if so
     if (isUriWhitelisted(request.getRequestURI, identityConfig.getWhiteList)) {
       logger.debug("Request URI matches a configured whitelist pattern! Allowing request to pass through.")
-      filterDirector.setFilterAction(FilterAction.PASS)
+      filterAction = FilterAction.PASS
     } else {
-      val requestHeaderManager = filterDirector.requestHeaderManager
-
       // Set the default behavior for this filter
-      filterDirector.setFilterAction(FilterAction.RETURN)
-      filterDirector.setResponseStatusCode(HttpServletResponse.SC_INTERNAL_SERVER_ERROR)
```

Comment:
Found another pattern that will break things. Setting these "default" values without overriding them. Specifically, if the response code is >= 500 when it reaches the end of the filter chain, the code that actually routes the request will simply return it instead.

So here's what changed. The FilterLogicHandlerDelegate would not apply the change to the response status until after calling the handleResponse method. In other words, it would delay applying the status to the underlying response until the last moment. With the new wrappers, that change is applied immediately.